File: server/routes/items.py
from flask import Blueprint, request, jsonify
from models import Item, Claim, User, db
from flask import current_app
import requests
import base64
from ai_models.ai_service import (
    analyze_image,
    generate_verification_question,
    find_matches_with_images,
)
import os
import json
import traceback
import base64
import jwt
from datetime import datetime
from werkzeug.utils import secure_filename
import cloudinary
import cloudinary.uploader
from routes.auth import SECRET_KEY
from PIL import Image
import io

items_bp = Blueprint("items", __name__)


@items_bp.route("/<int:item_id>/poster", methods=["GET"])
def generate_poster(item_id):
    """
    DEPRECATED: Poster generation is now handled client-side in Poster.jsx.
    This endpoint is left as a stub or can be removed entirely.
    """
    return jsonify({"message": "Use client-side poster generation at /item/<id>/poster"}), 410

# Cloudinary Configuration
cloudinary.config(
    cloud_name=os.getenv("CLOUDINARY_CLOUD_NAME"),
    api_key=os.getenv("CLOUDINARY_API_KEY"),
    api_secret=os.getenv("CLOUDINARY_API_SECRET"),
    secure=True,
)


@items_bp.route("/", methods=["POST"])
def create_item():
    """
    Report a new Lost or Found item.
    - Uploads image
    - Runs AI analysis (OpenAI Vision) to auto-tag (category, color, etc.)
    - Stores outcome in DB
    """

    # print("DEBUG: Entered create_item")
    try:
        # 1. Handle Image Upload
        if "image" not in request.files:
            return jsonify({"error": "No image uploaded"}), 400

        file = request.files["image"]
        if file.filename == "":
            return jsonify({"error": "No selected file"}), 400

        # ... (Processing Base64 and Image) ...
        # Process Image for DB (Base64)

        file_content = file.read()
        
        # --- Image Compression ---
        # Open image using Pillow
        img = Image.open(io.BytesIO(file_content))
        
        # Convert to RGB (in case of RGBA/PNG)
        if img.mode != 'RGB':
            img = img.convert('RGB')
            
        # Resize if too large (Max dimension 1024px)
        max_size = (1024, 1024)
        img.thumbnail(max_size, Image.Resampling.LANCZOS)
        
        # Save to buffer with compression
        compressed_buffer = io.BytesIO()
        img.save(compressed_buffer, format="JPEG", quality=70, optimize=True)
        compressed_buffer.seek(0)
        
        # Update file_content to use the compressed version
        compressed_content = compressed_buffer.getvalue()
        
        # Generate Base64 for AI (using compressed image is fine and faster)
        base64_data = base64.b64encode(compressed_content).decode("utf-8")
        mime_type = file.content_type or "image/jpeg"
        image_data_uri = f"data:{mime_type};base64,{base64_data}"

        # Local Backup for dev/debugging
        if os.getenv("VERCEL") or os.getenv("FLASK_ENV") == "production":
            upload_folder = os.path.join("/tmp", "uploads")
        else:
            upload_folder = os.path.join(
                os.path.dirname(os.path.dirname(__file__)), "uploads"
            )

        if not os.path.exists(upload_folder):
            os.makedirs(upload_folder)

        timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
        filename = f"{timestamp}_{secure_filename(file.filename)}"
        filepath = os.path.join(upload_folder, filename)

        with open(filepath, "wb") as f:
            f.write(file_content)

        # Cloudinary Upload
        try:
            print("DEBUG: Uploading to Cloudinary...")
            # We can upload the file_content (bytes) directly
            # Use compressed_content to save bandwidth/storage
            upload_result = cloudinary.uploader.upload(
                compressed_content, 
                folder="campusfind",
                resource_type="image"
            )
            image_url = upload_result.get("secure_url")
            print(f"DEBUG: Cloudinary Upload Success: {image_url}")
        except Exception as e:
            print(f"ERROR: Cloudinary Upload Failed: {e}")
            return jsonify({"error": f"Image upload failed: {str(e)}"}), 500

        # 2. AI Analysis (Auto-Tagging)
        print("DEBUG: Starting AI analysis...")
        try:
            data = request.form
            # Pass the data URI directly to avoid disk dependency issues on serverless
            analysis = analyze_image(image_data_uri, data.get("description", ""))
            print(f"DEBUG: AI Analysis result: {analysis}")
        except Exception as ai_e:
            print(f"DEBUG: AI Analysis Failed: {ai_e}")
            # Fallback to defaults if AI fails (e.g., quotas, network)
            analysis = {
                "category": "General Item",
                "color": "See image",
                "brand": None,
                "description": data.get("description", "Check image for details"),
                "distinctive_features": [],
            }

        # 2a. Handle Manual Tags (User Overrides)
        manual_tags_str = data.get("manual_tags", "[]")
        try:
            manual_tags = json.loads(manual_tags_str)
            if not isinstance(manual_tags, list):
                manual_tags = []
        except:
            manual_tags = []

        # Merge AI features and manual tags
        ai_features = analysis.get("distinctive_features", [])
        all_features = list(set(ai_features + manual_tags))
        analysis["distinctive_features"] = all_features

        # 3. Get User ID from Token
        token = request.headers.get("Authorization")
        if not token or not token.startswith("Bearer "):
            return jsonify({"error": "Unauthorized: Missing or invalid token"}), 401

        try:
            token_str = token.split(" ")[1]
            data = jwt.decode(token_str, SECRET_KEY, algorithms=["HS256"])
            user_id = data["user_id"]
        except Exception as e:
            return jsonify({"error": "Unauthorized: Invalid token"}), 401

        # 4. Create Item Record
        data = request.form

        final_category = (
            data.get("category") if data.get("category") else analysis.get("category")
        )
        final_color = data.get("color") if data.get("color") else analysis.get("color")
        final_brand = data.get("brand") if data.get("brand") else analysis.get("brand")

        new_item = Item(
            user_id=user_id,
            type=data.get("type", "lost"),
            description=analysis.get(
                "description", data.get("description", "No description")
            ),
            location=data.get("location", "Unknown"),
            date_lost=datetime.utcnow(),
            image_url=filename,
            image_data=image_url, # Store Cloudinary URL
            category=final_category,
            color=final_color,
            brand=final_brand,
            distinctive_features=json.dumps(analysis.get("distinctive_features", [])),
            contact_info=data.get("contact_info"),
        )

        # 5. Generate Verification Question if it's a 'Found' item
        # This helps the founder verify if a claimant is the true owner
        if new_item.type == "found":
            try:
                distinctive_features = analysis.get("distinctive_features", [])
                vq = generate_verification_question(
                    new_item.description, distinctive_features
                )
                new_item.verification_question = vq.get("question")
                new_item.verification_answer_type = vq.get("expected_answer_type")
            except Exception as vq_e:
                print(f"DEBUG: VQ Gen Failed: {vq_e}")

        # Gamification: Award 5 Points for Reporting (ONLY for FOUND items)
        # We reward people for helping others, not for losing things!
        if new_item.type == "found":
            try:

                user = User.query.get(user_id)
                if user:
                    user.trust_score = getattr(user, "trust_score", 0) + 5
            except Exception as xp_e:
                print(f"XP Update Failed: {xp_e}")

        db.session.add(new_item)
        db.session.commit()

        return (
            jsonify(
                {
                    "message": "Item reported successfully",
                    "item": {
                        "id": new_item.id,
                        "description": new_item.description,
                        "image_url": new_item.image_data,
                        "ai_tags": {
                            "category": new_item.category,
                            "color": new_item.color,
                            "brand": new_item.brand,
                        },
                    },
                }
            ),
            201,
        )

    except Exception as e:
        print(f"DEBUG: CRITICAL ERROR: {e}")
        traceback.print_exc()
        return jsonify({"error": str(e)}), 500


@items_bp.route("/", methods=["GET"])
def get_items():
    """
    Get main feed of items.
    Supports filtering by type (lost/found/all), search queries, and status.
    """
    try:
        type_filter = request.args.get("type")
        search_query = request.args.get("q")

        # Start query
        query = Item.query

        # FILTER: Exclude 'claimed' items from the public feed to keep it fresh
        # Unless specifically requested (e.g. for Stats page)
        if request.args.get("include_claimed") != "true" and type_filter != "all":
            query = query.filter(Item.status != "claimed")

        if type_filter and type_filter != "all":
            query = query.filter_by(type=type_filter)

        if search_query:
            search = f"%{search_query}%"
            query = query.filter(
                (Item.description.ilike(search))
                | (Item.category.ilike(search))
                | (Item.brand.ilike(search))
                | (Item.location.ilike(search))
            )

        # Sort by newest first
        items = query.order_by(Item.date_lost.desc()).all()

        result = []
        for item in items:
            img_src = item.image_data
            if not img_src and item.image_url:
                img_src = f"/uploads/{os.path.basename(item.image_url)}"

            result.append(
                {
                    "id": item.id,
                    "type": item.type,
                    "description": item.description,
                    "location": item.location,
                    "date_lost": item.date_lost.strftime("%Y-%m-%d %H:%M"),
                    "image_url": img_src,
                    "category": item.category,
                    "color": item.color,
                    "brand": item.brand,
                    "distinctive_features": (
                        json.loads(item.distinctive_features)
                        if item.distinctive_features
                        else []
                    ),
                    "status": item.status,
                }
            )

        return jsonify(result), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@items_bp.route("/my", methods=["GET"])
def get_my_items():
    token = request.headers.get("Authorization")
    if not token or not token.startswith("Bearer "):
        return jsonify({"error": "Unauthorized"}), 401

    try:

        # from models import Claim # Imported at top level now

        token = token.split(" ")[1]
        data = jwt.decode(token, SECRET_KEY, algorithms=["HS256"])
        user_id = data["user_id"]

        # 1. Items uploaded by me
        my_uploads = Item.query.filter_by(user_id=user_id).all()

        # 2. Items claimed by me (where I am the claimant/finder/respondent)
        my_claims = Claim.query.filter_by(claimant_id=user_id).all()
        my_claimed_items = [c.item for c in my_claims]

        # Combine unique items
        all_items_map = {}

        for item in my_uploads:
            all_items_map[item.id] = item

        for item in my_claimed_items:
            all_items_map[item.id] = item

        # Convert to list and sort
        combined_items = list(all_items_map.values())
        combined_items.sort(key=lambda x: x.date_lost, reverse=True)

        result = []
        for item in combined_items:
            img_src = item.image_data
            if not img_src and item.image_url:
                img_src = f"/uploads/{os.path.basename(item.image_url)}"

            result.append(
                {
                    "id": item.id,
                    "type": item.type,
                    "description": item.description,
                    "location": item.location,
                    "date_lost": item.date_lost.strftime("%Y-%m-%d %H:%M"),
                    "image_url": img_src,
                    "category": item.category,
                    "color": item.color,
                    "brand": item.brand,
                    "distinctive_features": (
                        json.loads(item.distinctive_features)
                        if item.distinctive_features
                        else []
                    ),
                    "status": item.status,
                }
            )
        return jsonify(result), 200
    except Exception as e:
        print(f"My Items Error: {e}")
        return jsonify({"error": str(e)}), 401


@items_bp.route("/<int:id>", methods=["GET"])
def get_item(id):
    """Get single item details"""
    try:
        item = Item.query.get_or_404(id)

        img_src = item.image_data
        if not img_src and item.image_url:
            img_src = f"/uploads/{os.path.basename(item.image_url)}"

        return (
            jsonify(
                {
                    "id": item.id,
                    "type": item.type,
                    "description": item.description,
                    "location": item.location,
                    "date_lost": item.date_lost.strftime("%Y-%m-%d %H:%M"),
                    "image_url": img_src,
                    "category": item.category,
                    "color": item.color,
                    "brand": item.brand,
                    "distinctive_features": (
                        json.loads(item.distinctive_features)
                        if item.distinctive_features
                        else []
                    ),
                    "status": item.status,
                    "user_id": item.user_id,
                    "reporter": {
                        "name": item.user.name,
                        "email": item.user.email,
                        "phone": getattr(item.user, "phone", None),
                        "profile_photo": getattr(item.user, "profile_photo", None),
                        "contact_info": item.contact_info,
                    },
                }
            ),
            200,
        )
    except Exception as e:
        return jsonify({"error": str(e)}), 404


@items_bp.route("/<int:id>/analyze", methods=["POST"])
def reanalyze_item(id):
    """
    Manually trigger AI analysis for an existing item.
    Useful for items that failed analysis or were uploaded before AI fixes.
    """
    try:

        item = Item.query.get_or_404(id)

        if not item.image_url:
            return jsonify({"error": "Item has no image"}), 400

        # Get full path
        img_path = os.path.join(
            os.path.dirname(os.path.dirname(__file__)),
            "uploads",
            os.path.basename(item.image_url),
        )

        if not os.path.exists(img_path):
            return jsonify({"error": "Image file not found on server"}), 404

        # Run Analysis
        analysis = analyze_image(img_path, item.description)

        # Update Item
        item.category = analysis.get("category")
        item.color = analysis.get("color")
        item.brand = analysis.get("brand")
        item.distinctive_features = json.dumps(analysis.get("distinctive_features", []))

        # Smart Description Update: Only update if AI gives a better description
        ai_desc = analysis.get("description")
        if ai_desc and ai_desc != "Check image for details":
            # We might keep user description but maybe append AI notes?
            # For now, let's trust the user's title unless it's generic
            pass

        db.session.commit()

        return (
            jsonify(
                {
                    "message": "Analysis updated",
                    "tags": {
                        "category": item.category,
                        "color": item.color,
                        "features": json.loads(item.distinctive_features),
                    },
                }
            ),
            200,
        )

    except Exception as e:
        print(f"Re-analysis failed: {e}")
        return jsonify({"error": str(e)}), 500


@items_bp.route("/match/<int:id>", methods=["GET"])
def get_matches(id):
    """
    🎯 FLAGSHIP FEATURE: Get AI-powered image matches for an item
    Compares the item against opposite-type items using Gemini Vision
    """
    try:

        # Get source item
        source_item = Item.query.get_or_404(id)

        # Get opposite type items that are unresolved
        opposite_type = "found" if source_item.type == "lost" else "lost"
        candidates = Item.query.filter_by(type=opposite_type, status="unresolved").all()

        if not candidates:
            return jsonify([]), 200

        matches = []

        # 1. Try AI Matching
        try:
            matches = find_matches_with_images(source_item, candidates)
        except Exception as ai_e:
            print(f"AI Match Failed (Rate Limit?): {ai_e}")
            matches = []  # Fallback

        # 2. Fallback: DB Pattern Matching if AI returned 0 matches
        if len(matches) == 0:
            print("Using DB Fallback for Matching...")
            fallback_matches = []
            for cand in candidates:
                score = 0
                reasons = []

                # Check Category
                if source_item.category and cand.category:
                    if source_item.category.lower() == cand.category.lower():
                        score += 40
                        reasons.append(f"Same category ({source_item.category})")

                # Check Color
                if source_item.color and cand.color:
                    if (
                        source_item.color.lower() in cand.color.lower()
                        or cand.color.lower() in source_item.color.lower()
                    ):
                        score += 30
                        reasons.append(f"Similar color ({source_item.color})")

                # Check Brand
                if source_item.brand and cand.brand:
                    if source_item.brand.lower() in cand.brand.lower():
                        score += 20
                        reasons.append(f"Same brand ({source_item.brand})")

                if score >= 30:  # Threshold
                    fallback_matches.append(
                        {
                            "item": {
                                "id": cand.id,
                                "description": cand.description,
                                "image_url": (
                                    cand.image_data
                                    if cand.image_data
                                    else (
                                        f"/uploads/{os.path.basename(cand.image_url)}"
                                        if cand.image_url
                                        else None
                                    )
                                ),
                            },
                            "confidence": score,
                            "reasoning": "Basic Feature Match: " + ", ".join(reasons),
                        }
                    )

            # Sort by score
            matches = sorted(
                fallback_matches, key=lambda x: x["confidence"], reverse=True
            )[:5]

        return jsonify(matches), 200
    except Exception as e:
        print(f"Error in get_matches: {e}")
        return jsonify({"error": str(e)}), 500

File: backend/routes/items.js
const express = require("express");
const multer = require("multer");
const cloudinary = require("cloudinary").v2;
// sharp is loaded lazily inside route handlers to avoid crashing on Vercel
const crypto = require("crypto");
const fs = require("fs/promises");
const jwt = require("jsonwebtoken");
const path = require("path");
const { Item, User, Claim } = require("../models");
const { SECRET_KEY } = require("../middleware/auth");
const { analyzeImage, generateVerificationQuestion } = require("../services/aiService");
const { Op } = require("sequelize");

const router = express.Router();
const LOCKING_CLAIM_STATUSES = ["accepted", "completed"];

// Multer: store in memory for processing
const upload = multer({ storage: multer.memoryStorage(), limits: { fileSize: 10 * 1024 * 1024 } });

// Cloudinary config
cloudinary.config({
  cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
  api_key: process.env.CLOUDINARY_API_KEY,
  api_secret: process.env.CLOUDINARY_API_SECRET,
  secure: true,
});

const localUploadDir = path.join(__dirname, "..", "uploads", "items");

const hasCloudinaryConfig = () =>
  Boolean(
    process.env.CLOUDINARY_CLOUD_NAME &&
      process.env.CLOUDINARY_API_KEY &&
      process.env.CLOUDINARY_API_SECRET
  );

const parseFeatures = (value) => {
  try {
    const parsed = JSON.parse(value || "[]");
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
};

const lockedClaimForItem = async (itemId) =>
  Claim.findOne({
    where: {
      item_id: itemId,
      status: { [Op.in]: LOCKING_CLAIM_STATUSES },
    },
    order: [
      ["status", "DESC"],
      ["timestamp", "DESC"],
    ],
  });

const claimStateForItem = async (item) => {
  if (item.status === "claimed") return "claimed";

  const lockedClaim = await lockedClaimForItem(item.id);
  if (!lockedClaim) return "available";

  return lockedClaim.status === "completed" ? "claimed" : "in_progress";
};

const textValue = (value) => (typeof value === "string" ? value.trim() : "");

const isBase64Image = (value) =>
  value.length > 100 && /^[A-Za-z0-9+/]+={0,2}$/.test(value);

const cloudinaryPathUrl = (value) => {
  const cloudName = textValue(process.env.CLOUDINARY_CLOUD_NAME);
  const normalized = textValue(value).replace(/^\/+/, "");

  if (!cloudName || !normalized) return "";
  if (normalized.startsWith("res.cloudinary.com/")) return `https://${normalized}`;
  if (normalized.startsWith("image/upload/")) {
    return `https://res.cloudinary.com/${cloudName}/${normalized}`;
  }

  const publicId =
    /^v\d+\//.test(normalized) || normalized.includes("/")
      ? normalized
      : `campusfind/${normalized}`;

  return `https://res.cloudinary.com/${cloudName}/image/upload/${publicId}`;
};

const directImageSource = (value) => {
  const normalized = textValue(value);

  if (!normalized) return "";
  if (
    normalized.startsWith("http://") ||
    normalized.startsWith("https://") ||
    normalized.startsWith("data:") ||
    normalized.startsWith("blob:")
  ) {
    return normalized;
  }
  if (normalized.startsWith("//res.cloudinary.com/")) return `https:${normalized}`;
  if (
    normalized.startsWith("res.cloudinary.com/") ||
    normalized.startsWith("image/upload/")
  ) {
    return cloudinaryPathUrl(normalized);
  }

  return "";
};

const storedImageDataSource = (value) => {
  const normalized = textValue(value);
  const directSource = directImageSource(normalized);

  if (directSource) return directSource;
  if (isBase64Image(normalized)) return `data:image/jpeg;base64,${normalized}`;
  return "";
};

const reconstructedCloudinaryUrl = (value) => {
  const normalized = textValue(value);

  if (!normalized || normalized.startsWith("/") || normalized.startsWith("uploads/")) {
    return "";
  }

  return directImageSource(normalized) || cloudinaryPathUrl(normalized);
};

const localUploadUrl = (value) => {
  const normalized = textValue(value);

  if (!normalized) return "";
  if (normalized.startsWith("/")) return normalized;
  if (normalized.startsWith("uploads/")) return `/${normalized}`;
  return `/uploads/${normalized}`;
};

const imageFromItem = (item) => {
  const imageUrl = textValue(item.image_url);
  const directImageUrl = directImageSource(imageUrl);
  const storedImage = storedImageDataSource(item.image_data);
  const cloudinaryUrl = reconstructedCloudinaryUrl(imageUrl);

  if (directImageUrl) return directImageUrl;
  if (storedImage) return storedImage;
  if (cloudinaryUrl) return cloudinaryUrl;
  if (imageUrl) return localUploadUrl(imageUrl);
  return "";
};

const saveImageLocally = async (buffer) => {
  await fs.mkdir(localUploadDir, { recursive: true });
  const filename = `${Date.now()}-${crypto.randomBytes(8).toString("hex")}.jpg`;
  await fs.writeFile(path.join(localUploadDir, filename), buffer);
  return `/uploads/items/${filename}`;
};

const uploadItemImage = async (buffer) => {
  if (!hasCloudinaryConfig()) {
    if (process.env.VERCEL) {
      throw new Error("Cloudinary credentials are required on Vercel");
    }
    return saveImageLocally(buffer);
  }

  try {
    const result = await new Promise((resolve, reject) => {
      const stream = cloudinary.uploader.upload_stream(
        { folder: "campusfind", resource_type: "image" },
        (error, result) => (error ? reject(error) : resolve(result))
      );
      stream.end(buffer);
    });

    return result.secure_url;
  } catch (error) {
    if (process.env.NODE_ENV === "production" || process.env.VERCEL) {
      throw error;
    }

    console.warn("Cloudinary upload failed. Saved image locally:", error.message);
    return saveImageLocally(buffer);
  }
};

const imageForAnalysis = async (item) => {
  if (item.image_data) return imageFromItem(item);
  const imageUrl = imageFromItem(item);
  if (imageUrl && imageUrl.startsWith("/uploads/")) {
    const uploadPath = imageUrl.replace(/^\/+/, "");
    const file = await fs.readFile(path.join(__dirname, "..", uploadPath));
    return `data:image/jpeg;base64,${file.toString("base64")}`;
  }
  return imageUrl;
};

const itemSummary = async (item) => {
  const claimState = await claimStateForItem(item);

  return {
    id: item.id,
    type: item.type,
    description: item.description,
    location: item.location,
    date_lost: item.date_lost ? item.date_lost.toISOString().slice(0, 16).replace("T", " ") : "",
    image_url: imageFromItem(item),
    category: item.category,
    color: item.color,
    brand: item.brand,
    distinctive_features: parseFeatures(item.distinctive_features),
    status: item.status,
    claim_state: claimState,
    claim_locked: claimState === "claimed",
  };
};

// ==================== Create Item ====================
router.post("/", upload.single("image"), async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({ error: "No image uploaded" });
    }

    const sharp = require("sharp");
    const compressed = await sharp(req.file.buffer)
      .resize(1024, 1024, { fit: "inside", withoutEnlargement: true })
      .jpeg({ quality: 70 })
      .toBuffer();

    let image_url = "";
    try {
      image_url = await uploadItemImage(compressed);
    } catch (uploadErr) {
      return res.status(500).json({ error: "Image upload failed: " + uploadErr.message });
    }

    const base64Data = compressed.toString("base64");
    const imageDataUri = `data:image/jpeg;base64,${base64Data}`;

    let analysis;
    try {
      analysis = await analyzeImage(imageDataUri, req.body.description || "");
    } catch (aiErr) {
      analysis = {
        category: "General Item",
        color: "See image",
        brand: null,
        description: req.body.description || "Check image for details",
        distinctive_features: [],
      };
    }

    let manualTags = [];
    try {
      manualTags = JSON.parse(req.body.manual_tags || "[]");
      if (!Array.isArray(manualTags)) manualTags = [];
    } catch { manualTags = []; }

    const aiFeatures = analysis.distinctive_features || [];
    const allFeatures = [...new Set([...aiFeatures, ...manualTags])];

    const header = req.headers.authorization;
    if (!header || !header.startsWith("Bearer ")) {
      return res.status(401).json({ error: "Unauthorized: Missing token" });
    }

    let userId;
    try {
      const token = header.split(" ")[1];
      const decoded = jwt.verify(token, SECRET_KEY);
      userId = decoded.userId;
    } catch {
      return res.status(401).json({ error: "Unauthorized: Invalid token" });
    }

    const finalCategory = req.body.category || analysis.category || "";
    const finalColor = req.body.color || analysis.color || "";
    const finalBrand = req.body.brand || analysis.brand || "";

    const newItemData = {
      user_id: userId,
      type: req.body.type || "lost",
      description: analysis.description || req.body.description || "No description",
      location: req.body.location || "Unknown",
      date_lost: new Date(),
      image_url,
      category: finalCategory,
      color: finalColor,
      brand: finalBrand,
      distinctive_features: JSON.stringify(allFeatures),
      contact_info: req.body.contact_info || "",
    };

    let verification_question = "";
    let verification_answer_type = "";

    if (newItemData.type === "found") {
      try {
        const vq = await generateVerificationQuestion(newItemData.description, allFeatures);
        verification_question = vq.question || "";
        verification_answer_type = vq.expected_answer_type || "";
      } catch (vqErr) {
        console.warn("Verification question generation failed:", vqErr.message);
      }
      try {
        const u = await User.findByPk(userId);
        if (u) {
          u.trust_score = (u.trust_score || 0) + 5;
          await u.save();
        }
      } catch (xpErr) {}
    }

    newItemData.verification_question = verification_question;
    newItemData.verification_answer_type = verification_answer_type;

    const newItem = await Item.create(newItemData);

    res.status(201).json({
      message: "Item reported successfully",
      item: {
        id: newItem.id,
        description: newItem.description,
        image_url: newItem.image_url,
        ai_tags: {
          category: newItem.category,
          color: newItem.color,
          brand: newItem.brand,
        },
      },
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// ==================== Get Items (Feed) ====================
router.get("/", async (req, res) => {
  try {
    const { type, q, include_claimed } = req.query;
    const where = {};

    if (include_claimed !== "true") {
      where.status = { [Op.ne]: "claimed" };
    }

    if (type && type !== "all") {
      where.type = type;
    }

    if (q) {
      where[Op.or] = [
        { description: { [Op.iLike]: `%${q}%` } },
        { category: { [Op.iLike]: `%${q}%` } },
        { brand: { [Op.iLike]: `%${q}%` } },
        { location: { [Op.iLike]: `%${q}%` } },
      ];
    }

    const items = await Item.findAll({ 
      where, 
      order: [['date_lost', 'DESC']] 
    });

    const summaries = await Promise.all(items.map(itemSummary));
    const result =
      include_claimed === "true"
        ? summaries
        : summaries.filter((item) => item.claim_state !== "claimed");

    res.json(result);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// ==================== Get My Items ====================
router.get("/my", async (req, res) => {
  const header = req.headers.authorization;
  if (!header || !header.startsWith("Bearer ")) {
    return res.status(401).json({ error: "Unauthorized" });
  }

  try {
    const token = header.split(" ")[1];
    const decoded = jwt.verify(token, SECRET_KEY);
    const userId = decoded.userId;

    const myUploads = await Item.findAll({ where: { user_id: userId } });

    const myClaims = await Claim.findAll({ 
      where: { claimant_id: userId },
      include: [{ model: Item }] 
    });
    
    const claimedItems = myClaims.map((c) => c.Item).filter(Boolean);

    const allItemsMap = {};
    myUploads.forEach((item) => (allItemsMap[item.id.toString()] = item));
    claimedItems.forEach((item) => (allItemsMap[item.id.toString()] = item));

    const combined = Object.values(allItemsMap).sort(
      (a, b) => new Date(b.date_lost) - new Date(a.date_lost)
    );

    const result = await Promise.all(combined.map(itemSummary));

    res.json(result);
  } catch (error) {
    res.status(401).json({ error: error.message });
  }
});

// ==================== Get Single Item ====================
router.get("/:id", async (req, res) => {
  try {
    const item = await Item.findByPk(req.params.id, {
      include: [{ model: User }]
    });
    if (!item) return res.status(404).json({ error: "Item not found" });

    const reporter = item.User;

    const summary = await itemSummary(item);

    res.json({
      ...summary,
      user_id: reporter.id,
      reporter: {
        name: reporter.name,
        email: reporter.email,
        phone: reporter.phone || null,
        profile_photo: reporter.profile_photo || null,
        contact_info: item.contact_info,
      },
    });
  } catch (error) {
    res.status(404).json({ error: error.message });
  }
});

// ==================== Re-analyze Item (AI) ====================
router.post("/:id/analyze", async (req, res) => {
  try {
    const item = await Item.findByPk(req.params.id);
    if (!item) return res.status(404).json({ error: "Item not found" });
    if (!imageFromItem(item)) return res.status(400).json({ error: "Item has no image" });

    const analysis = await analyzeImage(await imageForAnalysis(item), item.description);

    item.category = analysis.category || item.category;
    item.color = analysis.color || item.color;
    item.brand = analysis.brand || item.brand;
    item.distinctive_features = JSON.stringify(
      analysis.distinctive_features || parseFeatures(item.distinctive_features)
    );
    await item.save();

    res.json({
      message: "Analysis updated",
      tags: {
        category: item.category,
        color: item.color,
        features: parseFeatures(item.distinctive_features),
      },
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// ==================== Get Matches (PostgreSQL Fallback) ====================
router.get("/match/:id", async (req, res) => {
  try {
    const sourceItem = await Item.findByPk(req.params.id);
    if (!sourceItem) return res.status(404).json({ error: "Item not found" });

    const oppositeType = sourceItem.type === "lost" ? "found" : "lost";
    let matches = [];

    // Fallback matching uses simple tags and keywords.
    const terms = [sourceItem.category, sourceItem.brand, sourceItem.color].filter(Boolean);
    const features = parseFeatures(sourceItem.distinctive_features);
    const allTerms = [...terms, ...features];

    if (allTerms.length > 0) {
      const orConditions = allTerms.map(term => ({
        [Op.or]: [
          { description: { [Op.iLike]: `%${term}%` } },
          { category: { [Op.iLike]: `%${term}%` } },
          { distinctive_features: { [Op.iLike]: `%${term}%` } }
        ]
      }));

      const candidates = await Item.findAll({
        where: {
          type: oppositeType,
          status: "unresolved",
          [Op.or]: orConditions
        },
        limit: 20
      });

      const candidateSummaries = await Promise.all(candidates.map(itemSummary));
      const availableCandidates = candidateSummaries.filter((cand) => !cand.claim_locked);

      // Simple scoring based on how many terms hit
      matches = availableCandidates.map(cand => {
        let hitCount = 0;
        const candText = `${cand.description} ${cand.category} ${cand.distinctive_features}`.toLowerCase();
        allTerms.forEach(term => {
          if (candText.includes(term.toLowerCase())) hitCount++;
        });
        
        return {
          id: cand.id,
          confidence: Math.min(Math.round((hitCount / allTerms.length) * 100), 95),
          reasoning: "Keyword Text Match",
          item: {
            id: cand.id,
            description: cand.description,
            image_url: cand.image_url,
            category: cand.category,
            color: cand.color,
            claim_state: cand.claim_state,
            claim_locked: cand.claim_locked,
          },
        };
      }).filter(m => m.confidence > 20);

      matches.sort((a, b) => b.confidence - a.confidence);
    }

    res.json(matches.slice(0, 5));
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;
